client/src/common/client.ts

```diff
@@ -691,7 +709,45 @@ export abstract class BaseLanguageClient implements FeatureClient<Middleware, La
 		if (this._didChangeTextDocumentFeature!.syncKind === TextDocumentSyncKind.Full) {
 			await this.sendPendingFullTextDocumentChanges(connection);
 		}
-		return connection.sendRequest<R>(type, ...params);
+
+		const _sendRequest = this._clientOptions.middleware?.sendRequest;
+		if (_sendRequest !== undefined) {
+			let param: any | undefined = undefined;
+			let token: CancellationToken | undefined = undefined;
+
+			// Separate cancellation tokens from other parameters for a better client interface
+			if (params.length === 1) {
+				// CancellationToken is an interface, so we need to check if the first param complies to it
+				if ('onCancellationRequested' in params[0] && 'isCancellationRequested' in params[0]) {
```

Comment:
In general we do these checks with TS type check functions. IMO you should be able to simply use: https://insiders.vscode.dev/github/microsoft/vscode-languageserver-node/blob/fca08a7e9f9bd9beeb687a5ae5529139783e915b/jsonrpc/src/common/cancellation.ts#L43

Reply:
I imported and used it instead. Let me know if you believe this is indeed the way to go and if we should do the same for notifications.